src/Models/Message.php

```diff
@@ -89,4 +156,29 @@ protected static function newFactory()
     {
         return MessageFactory::new();
     }
+
+    /**
+     * Create a new Eloquent Collection instance.
+     *
+     * @param  array  $models
+     * @return \Illuminate\Database\Eloquent\Collection
+     */
+    public function newCollection(array $models = [])
+    {
+        $collection = new MessageCollection($models);
+
+        // If there is not authenticated user we can return the collection
+        // immediately. This because we can't mark the messages as read.
+        // Usually, this will only be the case while performing unit tests.
+        if (Auth::guest()) {
+            return $collection;
+        }
+
+        $markAsRead = filter_var(
+            request()->get('mark_messages_as_read', false),
+            FILTER_VALIDATE_BOOLEAN
+        );
+
+        return $collection->markAsRead(Auth::user(), $markAsRead);
```

Comment:
In my opinion this is totally the wrong place to do this. I would have put this in controller instead.

Reply:
That would generate duplicate code. Because currently this code is executed in the controllers and in the future maybe more. I tried to find a place where I only had to write the code once, that is here. I could make a manager or something to place the code in.

Reply:
What if we want to fetch some messages during a request without marking them as read, but the query parameter is set on true? This will automatically mark them as read. I know it is very unlikely but this could result some very annoying to find bugs. What happens during eagerloading for example?

Still think this is a route specific feature.